Feature/14706 epilogue screen for blogging reminders

ravishanker opened the pull request:
Fixes #14706 

This builds off this [work in progress](https://github.com/wordpress-mobile/WordPress-Android/pull/14753) by @planarvoid.

Only need to review this [commit](https://github.com/wordpress-mobile/WordPress-Android/pull/14768/commits/2f2cc26d4b87080cada8769f4b29a44800e2741c)  here

<img width=320 src=https://user-images.githubusercontent.com/990349/122340022-1aabd180-cf85-11eb-8641-e5bfd7f189c2.png>

To test:

- Turn on the blogging reminders flag (Me/App settings/Test feature configuration - BloggingRemindersFeatureConfig)
- Create a new post
- Notice the blogging reminders prologue screen is shown
- Click on "Set goals"
- Select days screen is shown, select days, then "Notify me" button is enabled
- Click on "Notify me" button
- Notice the epilogue screen is shown.  Done button is enabled, Clicking Done button dismisses bottomsheet

Alternatively, you can get to it from

- My Site > Settings > Blogging goals then selecting days on bottom sheet


## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
NA

3. What automated tests I added (or what prevented me from doing so)
Unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
Build epilogue screen

epilogue screen for blogging reminders

ravishanker committed:
Update image

use the right image for epilogue screen

ravishanker committed:
Update epilogue body text logic

Updated logic for body text on epilogue screen

ravishanker committed:
Update BloggingRemindersViewModelTest.kt

Fix unit test

ravishanker committed:
fix ktlint and detekt issues

renanferrari commented:
@ravishanker FYI I've changed the base branch to use the one in #14753 to make this a bit easier to review.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewModel.kt:
```diff
@@ -41,14 +45,14 @@ class BloggingRemindersViewModel @Inject constructor(
     private val _selectedScreen = MutableLiveData<Screen>()
     private val _bloggingRemindersModel = MutableLiveData<BloggingRemindersModel>()
-    val uiState: LiveData<List<BloggingRemindersItem>> = merge(
+    @ExperimentalStdlibApi val uiState: LiveData<List<BloggingRemindersItem>> = merge(
```
We were actually using [`LiveDataUtils.merge`](https://github.com/wordpress-mobile/WordPress-Android/blob/57c5b15a3d5d90cbdd8c1803f359a920a670033e/WordPress/src/main/java/org/wordpress/android/util/LiveDataUtils.kt#L93-L108) here instead of the experimental Kotlin API, so we shouldn't need to use this annotation.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewModel.kt:
```diff
@@ -101,9 +105,40 @@ class BloggingRemindersViewModel @Inject constructor(
+        val selectedDays = when {
+            numberOfDays <= THREE_DAYS -> {
+                bloggingRemindersModel?.enabledDays?.map { day ->
+                    day.name.toLowerCase(Locale.ROOT).replaceFirstChar {
```
Is there any reason why we are using `Locale.ROOT` here instead of `Locale.getDefault()`?

renanferrari replied:
Using the enum name here is not going to work for languages other than English.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewModel.kt:
```diff
@@ -101,9 +105,40 @@ class BloggingRemindersViewModel @Inject constructor(
+            numberOfDays <= THREE_DAYS -> {
+                bloggingRemindersModel?.enabledDays?.map { day ->
+                    day.name.toLowerCase(Locale.ROOT).replaceFirstChar {
+                        it.toUpperCase()
```
Instead of using `toUpperCase` or `toLowerCase` here, we should actually use the proper formatting classes to get the localized day of week name.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewModel.kt:
```diff
@@ -101,9 +105,40 @@ class BloggingRemindersViewModel @Inject constructor(
+                bloggingRemindersModel?.enabledDays?.map { day ->
+                    day.name.toLowerCase(Locale.ROOT).replaceFirstChar {
+                        it.toUpperCase()
+                    }.plus(" and ")
```
This won't work for languages other than English. The `and` should probably be included as part of the String resource, so it can be properly localized.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewModel.kt:
```diff
@@ -101,9 +105,40 @@ class BloggingRemindersViewModel @Inject constructor(
+                bloggingRemindersModel?.enabledDays?.map { day ->
+                    day.name.toLowerCase(Locale.ROOT).replaceFirstChar {
+                        it.toUpperCase()
+                    }.take(FIRST_THREE_CHARS).plus(", ")
```
Same as above, instead of using `take()` here, we should actually use the proper formatting classes to get the localized day of week short name.

I believe we can do this with `DateFormat` although I don't remember the syntax pretty well. That said, when #14698 gets merged, we will be able to do it like so: `DayOfWeek.MONDAY.getDisplayName(TextStyle.SHORT, Locale.getDefault())`.

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewModel.kt:
```diff
@@ -101,9 +105,40 @@ class BloggingRemindersViewModel @Inject constructor(
+                    }.take(FIRST_THREE_CHARS).plus(", ")
+                }.toString()
+            }
+            else -> "everyday"
```
We can probably merge this `when` block with the `when` block below where we return the actual text and then we wouldn't need to have this `else` branch here.

renanferrari reviewed (changes requested):
Thanks for the changes!

I actually couldn't make this work. I never saw the Epilogue screen, only this instead:

<img src="https://user-images.githubusercontent.com/830056/120876825-6c1d8f00-c589-11eb-9df6-b081387f889f.png" width=320/>

I haven't looked too deeply into the code to find out why, but I did see some issues with it, as I've commented below.

ravishanker commented:
> Thanks for the changes!
> 
> I actually couldn't make this work. I never saw the Epilogue screen, only this instead:
> 
> <img alt="" width="320/" src="https://user-images.githubusercontent.com/830056/120876825-6c1d8f00-c589-11eb-9df6-b081387f889f.png">
> 
> I haven't looked too deeply into the code to find out why, but I did see some issues with it, as I've commented below.

Thanks @renanferrari The button gets enabled only when the day selection is complete (I did with local manipulation to build this).  This needs to wait till all Vojta's PRs in train are merged.  
I should have mentioned before.  Also, all your comments are valid observations as it stands.  The logic needs updated once your PR and Vojta's PRs get merged as this is the last screen in that flow.

renanferrari commented:
@ravishanker Got it! I had a feeling something was missing, so that makes more sense 😄

AliSoftware commented:
Moving this PR to next milestone, since 17.6 goes into code freeze today.

ravishanker committed:
Merge branch 'develop' into feature/14706-epilogue-screen-for-blogging-reminders

renanferrari commented:
Hey @ravishanker, would you be able to merge the changes from #14827 on this PR so we can review the Epilogue screen with the entire flow? (and then also change the base branch from `develop` to `feature/14705-day-selection-screen-part-two`)

Edit: just realized I had linked the wrong PR above.

ravishanker committed:
Merge branch 'develop' into feature/14706-epilogue-screen-for-blogging-reminders

ravishanker committed:
Update BloggingRemindersViewModelTest.kt

fix test

ravishanker committed:
Merge branch 'feature/14705-day-selection-screen-part-two' into feature/14706-epilogue-screen-for-blogging-reminders

ravishanker committed:
Update epilogue screen logic

Updated epilogue screen number of times and selected days logic based merged PRs

ravishanker commented:
> Hey @ravishanker, would you be able to merge the changes from #14827 on this PR so we can review the Epilogue screen with the entire flow? (and then also change the base branch from `develop` to `feature/14705-day-selection-screen-part-two`)
> 
> Edit: just realized I had linked the wrong PR above.

Hey @renanferrari, I've merged @planarvoid PR #14827, resolved conflicts and cleaned up some logic for this one.  Needs one more pass, may be after your PRs are merged.

renanferrari committed:
Merge branch 'feature/14705-day-selection-screen-part-two' into feature/14706-epilogue-screen-for-blogging-reminders

renanferrari commented:
@ravishanker Just a heads up that I have merged `develop` into #14827, changed the base branch here to point to that and then merged that into this, so that the diff would be correctly recalculated. I tried doing that in a separate, intermediary branch, but that wasn't working for some reason (I suspect it's related to [the way GitHub handles this](https://github.com/isaacs/github/issues/750)). Hope you don't mind!

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/bloggingreminders/BloggingReminderBottomSheetFragment.kt:
```diff
@@ -31,6 +31,7 @@ class BloggingReminderBottomSheetFragment : BottomSheetDialogFragment() {
         return inflater.inflate(R.layout.recycler_view_primary_button_bottom_sheet, container)
     }
 
+    @ExperimentalStdlibApi
```
I believe we can remove this now?

renanferrari commented on WordPress/src/test/java/org/wordpress/android/ui/bloggingreminders/BloggingRemindersViewModelTest.kt:
```diff
@@ -50,6 +50,7 @@ class BloggingRemindersViewModelTest : BaseUnitTest() {
     private lateinit var events: MutableList<Boolean>
     private lateinit var uiState: MutableList<UiState>
 
+    @ExperimentalStdlibApi
```
Same as above.

renanferrari commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -3650,6 +3650,10 @@
     <string name="blogging_reminders_update">Update</string>
     <string name="blogging_reminders_done">Done</string>
+    <string name="blogging_reminders_epilogue_title">All set!</string>
+    <string name="blogging_reminders_epilogue_body_days">You\'ll get reminders to blog %1$s on %2$s.</string>
```
Instead of reusing the `R.array.blogging_goals_count` array to compose this string, we should probably use numbers directly, as it is defined in our design spec:

<img width="360" alt="Success sheet" src="https://user-images.githubusercontent.com/830056/122465708-cc9bcb80-cf8e-11eb-92e8-01bf4e7a66d2.png">

I know you are probably going to handle this in an upcoming PR, but I just want to make sure we're on the same page.

ravishanker replied:
I had it before like that **2** times.  

I've updated it to be consistent with the summary we show now in Site settings.  Also, it reads better For e.g. **1** time a week needs to read **once** a week or **2** times a week as **twice** a week and the rest as like so three times or four times...

Yeah, if the design review suggests to go with earlier design then I'll revert it back.

renanferrari reviewed (commented):
Thanks for the changes! I have tested the app and can confirm this is working as expected. Code is also looking good overall, I just left a few minor comments below.

As for the visual changes, they are looking good to me, but we should ask for a proper design review once you submit another PR to handle the selected days string formatting.

ravishanker committed:
Remove annotations

Remove redundant annotations

ravishanker committed:
Merge branch 'feature/14706-epilogue-screen-for-blogging-reminders' of github.com:wordpress-mobile/WordPress-Android into feature/14706-epilogue-screen-for-blogging-reminders

ravishanker committed:
Format body

format body and caption on epilogue screen

renanferrari commented:
Hey @ravishanker, let me know when this is ready for another round!

ravishanker committed:
Merge branch 'develop' into feature/14706-epilogue-screen-for-blogging-reminders

ravishanker committed:
Update BloggingRemindersViewModel.kt

fix lint error

ravishanker committed:
Update BloggingRemindersViewModel.kt

fix a lint error

ravishanker committed:
Update BloggingRemindersViewModel.kt

fix lint error

ravishanker commented:
> Hey @ravishanker, let me know when this is ready for another round!

Hi @renanferrari  please merge this, since Vojta merged his selection screen part-2.  I'll cover copy and formatting on this in the next PR.  Thanks

renanferrari reviewed (approved):
LGTM 👍
